File: frontend/src/shared/api/mocks/achievements.ts
import { type Achievement } from '../../types';

export const mockAchievements: Achievement[] = [
  { id: 'ach-1', name: 'Первый шаг', description: 'Зарегистрируйся на платформе', icon: '🎯', rarity: 'common', unlockedAt: '2025-09-01' },
  { id: 'ach-2', name: 'Автор', description: 'Загрузи свой первый материал', icon: '✍️', rarity: 'common', unlockedAt: '2025-09-05' },
  { id: 'ach-3', name: 'Популярный автор', description: '10+ покупок твоего материала', icon: '🔥', rarity: 'rare', unlockedAt: '2025-10-01' },
  { id: 'ach-4', name: 'Мега-эксперт', description: 'Загрузи 10+ материалов', icon: '🏆', rarity: 'epic', unlockedAt: '2025-11-15' },
  { id: 'ach-5', name: 'Ценный кадр', description: 'Рейтинг 4.8+ при 20+ оценках', icon: '💎', rarity: 'epic', unlockedAt: '2026-01-01' },
  { id: 'ach-6', name: 'Сообщество', description: 'Вступи в 3 сообщества', icon: '🤝', rarity: 'common' },
  { id: 'ach-7', name: 'Автор месяца', description: 'Стань топ-1 автором недели', icon: '👑', rarity: 'legendary', unlockedAt: '2026-02-01' },
  { id: 'ach-8', name: 'Легенда', description: '1000+ CodeCoins заработано', icon: '🌟', rarity: 'legendary', unlockedAt: '2026-03-01' },
  { id: 'ach-9', name: 'Челленджер', description: 'Выиграй челлендж сообщества', icon: '⚡', rarity: 'rare' },
  { id: 'ach-10', name: 'Книжный червь', description: 'Купи 50+ материалов', icon: '📚', rarity: 'epic' },
];

File: api/routers/achievements.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from database import get_db
from models.user import User
from models.achievement import Achievement, UserAchievement
from schemas.achievement import AchievementResponse
from services.auth import get_current_user

router = APIRouter(prefix="/achievements", tags=["achievements"])


@router.get("/", response_model=list[AchievementResponse])
async def get_all_achievements(db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Achievement))
    return [AchievementResponse.model_validate(a) for a in result.scalars().all()]


@router.get("/my", response_model=list[AchievementResponse])
async def get_my_achievements(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Achievement)
        .join(UserAchievement)
        .where(UserAchievement.user_id == current_user.id)
    )
    achievements = result.scalars().all()
    # Get unlock dates
    ua_result = await db.execute(
        select(UserAchievement).where(UserAchievement.user_id == current_user.id)
    )
    ua_map = {ua.achievement_id: ua.unlocked_at for ua in ua_result.scalars().all()}

    return [
        AchievementResponse(
            id=a.id,
            name=a.name,
            description=a.description,
            icon=a.icon,
            rarity=a.rarity,
            unlocked_at=ua_map.get(a.id),
        )
        for a in achievements
    ]


@router.get("/user/{user_id}", response_model=list[AchievementResponse])
async def get_user_achievements(user_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(Achievement)
        .join(UserAchievement)
        .where(UserAchievement.user_id == user_id)
    )
    achievements = result.scalars().all()
    ua_result = await db.execute(
        select(UserAchievement).where(UserAchievement.user_id == user_id)
    )
    ua_map = {ua.achievement_id: ua.unlocked_at for ua in ua_result.scalars().all()}

    return [
        AchievementResponse(
            id=a.id,
            name=a.name,
            description=a.description,
            icon=a.icon,
            rarity=a.rarity,
            unlocked_at=ua_map.get(a.id),
        )
        for a in achievements
    ]
